### PR TITLE
Fixed /restore and /graph conjunction error on exit code

### PR DESCRIPTION
**Pull request by maridematte**
Fixes https://github.com/dotnet/msbuild/issues/9443

### Context
When the `/graph` and `/restore` are used within the same command, the build exit code will always be 0. This is because of the variable used to define success of the restore action overrides the success of the graph build.

### Changes Made
Added an extra condition when defining the success of the build to account for this case.

### Testing
Made sure existing tests passed and added a unit test for this case, an some manual testing.

### CODE REVIEW

**Review by ladipro — APPROVED** [review]
nit: I think it would be slightly better if the `success =` assignments were done right after the respective results are calculated. The current code effectively repeats the condition `graphResult != null && !saveProjectResult` in two places in an non-obvious way and it would be easy to have them diverge and introduce a similar bug in the future. Especially given what a monster of a method this is.

File: src/MSBuild.UnitTests/XMake_Tests.cs [hunk 1]
```diff
@@ -627,6 +627,7 @@ public void InvalidMaxCPUCountSwitch4()
         [InlineData("-getProperty:Foo;Bar -getItem:MyItem -t:Build", true, "TargetValue", true, true, false, true, false)]
         [InlineData("-getProperty:Foo;Bar -getItem:MyItem", true, "EvalValue", true, false, false, true, false)]
         [InlineData("-getProperty:Foo;Bar -getTargetResult:MyTarget", true, "TargetValue", false, false, true, true, false)]
+        [InlineData("-getProperty:Foo;Bar -getTargetResult:MyTarget -t:restore", true, "TargetValue", false, false, true, true, false)]
```
Comment (rainersigwald) [inline src/MSBuild.UnitTests/XMake_Tests.cs:630]: I don't think I understand how this is related to the `-restore -graph` description of the PR.

File: src/MSBuild/XMake.cs [hunk 2]
```diff
@@ -1567,6 +1567,7 @@ private static void ResetGatheringSwitchesState()
                                             nodeResultKvp.Key.ProjectInstance.GlobalProperties.All(propertyKvp => entryPoint.GlobalProperties.TryGetValue(propertyKvp.Key, out string entryValue) &&
                                                                                                                                         entryValue.Equals(propertyKvp.Value)))
                                             .Value;
+                                        success = result.OverallResult == BuildResultCode.Success;
```
Comment (rainersigwald) [inline src/MSBuild/XMake.cs:1570]: `result`, or should this use `graphResult` like the not-saving-results version and let the `result`-filtering stuff be only for extracting saved results?

**Review by rainersigwald — COMMENTED** [review]
I like the refactoring a lot!

**Comment by AR-May** [comment]
I think we should target vs17.9 branch for this PR, since this is a bug?

File: src/MSBuild.UnitTests/XMake_Tests.cs [hunk 3]
```diff
@@ -697,6 +697,36 @@ public void InvalidMaxCPUCountSwitch4()
...
+            Console.WriteLine(""Hello, World!"")
+            A Line here for this to not compile right");
+
+            string graph = isGraphBuild ? "--graph" : "";
+            string result = RunnerUtilities.ExecMSBuild($" {project.Path} /restore {graph}", out bool success);
+
+            success.ShouldBeFalse();
+            result.Contains("Program.cs(2,47): error CS1002: ; expected");
+            result.Contains("Program.cs(3,20): error CS1003: Syntax error, ','");
+            result.Contains("Program.cs(3,54): error CS1002: ; expected");
```
Comment (rainersigwald) [inline src/MSBuild.UnitTests/XMake_Tests.cs:727]:
```suggestion
            result.ShouldContain("Program.cs(2,47): error CS1002: ; expected");
            result.ShouldContain("Program.cs(3,20): error CS1003: Syntax error, ','");
            result.ShouldContain("Program.cs(3,54): error CS1002: ; expected");
```

?

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**